lms/djangoapps/course_api/blocks/transformers/tests/test_milestones.py

```diff
@@ -0,0 +1,234 @@
+                ],
+            },
+            {
+                '#type': 'sequential',
+                '#ref': 'TimedExam',
+                'is_time_limited': True,
+                'is_proctored_enabled': False,
+                'is_practice_exam': False,
+                '#children': [
+                    {'#type': 'vertical', '#ref': 'J'},
+                    {'#type': 'vertical', '#ref': 'K'},
+                ],
+            },
+        ]
+
+    def test_special_exams_not_visible_to_non_staff(self):
+        self.get_blocks_and_check_against_expected(self.user, self.ALL_BLOCKS_EXCEPT_SPECIAL)
+
+    @ddt.data(
+        (
+            'H',
+            'A',
+            'B',
+            ('course', 'A', 'B', 'C',)
+        ),
+        (
+            'H',
+            'ProctoredExam',
+            'D',
+            ('course', 'A', 'B', 'C'),
+        ),
+    )
+    @ddt.unpack
+    def test_gated(self, gated_block_ref, gating_block_ref, gating_block_child, expected_blocks_before_completion):
+        """
+        First, checks that a student cannot see the gated block when it is gated by the gating block and no
+        attempt has been made to complete the gating block.
+        Then, checks that the student can see the gated block after the gating block has been completed.
+
+        expected_blocks_before_completion is the set of blocks we expect to be visible to the student
+        before the student has completed the gating block.
+
+        The test data includes one special exam and one non-special block as the gating blocks.
+        """
+        self.course.enable_subsection_gating = True
+        self.setup_gated_section(self.blocks[gated_block_ref], self.blocks[gating_block_ref])
+        self.get_blocks_and_check_against_expected(self.user, expected_blocks_before_completion)
+
+        # mock the api that the lms gating api calls to get the score for each block to always return 1 (ie 100%)
+        courseware.grades.get_module_score = Mock(return_value=1)
+
+        # this call triggers reevaluation of prerequisites fulfilled by the parent of the
+        # block passed in, so we pass in a child of the gating block
+        lms_gating_api.evaluate_prerequisite(
+            self.course,
+            UsageKey.from_string(unicode(self.blocks[gating_block_child].location)),
+            self.user.id)
+
+        self.get_blocks_and_check_against_expected(self.user, self.ALL_BLOCKS_EXCEPT_SPECIAL)
+
+    def test_staff_access_proctored(self):
+        """
+        Ensures that staff can always access all blocks in the course,
+        regardless of gating or proctoring.
+        """
+        expected_blocks = self.ALL_BLOCKS
+        self.setup_proctored_exam(
+            self.blocks['ProctoredExam'],
+            ProctoredExamStudentAttemptStatus.rejected,
+            self.user.id
+        )
+        self.setup_gated_section(self.blocks['PracticeExam'], self.blocks['ProctoredExam'])
```

Comment:
To decouple testing of gating from exams, setup the gated sections using Subsection A and Subsection H instead of using the exams.